Initial FIELD API and gpu-scc-field variant

mlange05 opened the pull request:
_Note: This PR precedes #25 and aims to bring in the base variant of the FIELD API. I'm filing it now as a standalone PR to ease review and integration of changes when we eventually switch to the new long-term supported standalone repo for FIELD API._

This PR revives an old experimentation with optimising data transfer costs via host-side memory pinning behind a common Fortran API for field objects. It therefor ships an early version of the so-called FIELD API (`field_module.F90`) that supports explicit memory pinning for individual field objects, and a bulk-allocation scheme similar to the state array handling in IFS. 

The PR adds a new variant `dwarf-cloudsc-gpu-scc-field` that uses the GPU-SCC kernel for convenience, but allocates field object via the new `CLOUDSC_FIELD_STATE` object. The driver-side FIELD API usage in demonstrated in the new kernel, and the default data transfer times should be superior on Nvidia GPUs to other GPU-SCC variants.

An environment flag to trigger bulk-allocation of arrays via FIELD API wrapper constructors is also provided, and can be triggered via `CLOUDSC_PACKED_STORAGE=ON`, as described in the README. This closely mimics IFS allocation behaviour, but reduces the data transfer performance due to the strided memory layout (thanks @awnawab for showing me this!).

reuterbal reviewed (approved):
This is great and I very much appreciate this digestable stepping stone towards #25. Very neatly documented and integrated. I've only left a few minor comments, nothing that stops this from going in.

awnawab commented on arch/toolchains/ecmwf-hpc2020-nvhpc.cmake:
```diff
@@ -50,6 +50,8 @@ set(ECBUILD_Fortran_FLAGS "${ECBUILD_Fortran_FLAGS} -Ktrap=fp")
 set(ECBUILD_Fortran_FLAGS "${ECBUILD_Fortran_FLAGS} -Kieee")
 set(ECBUILD_Fortran_FLAGS "${ECBUILD_Fortran_FLAGS} -Mdaz")
 
+set(ECBUILD_Fortran_LINK_FLAGS "-gpu=pinned")
```
Yes I agree, we just need to add this as a link option for the `dwarf-cloudsc-gpu-scc-field` executable, behind a `HAVE_CUDA` test.

awnawab reviewed (approved):
As @reuterbal said, very cleanly done and nicely documented. It also makes the focus of #25 a lot clearer and makes it much easier to integrate; thanks for that! I just mirrored two of B's comments on the cmake side, neither of which is a showstopper by any means, so this can go in as is.

mlange05 commented on src/cloudsc_gpu/dwarf_cloudsc_gpu.F90:
```diff
@@ -33,17 +33,29 @@ PROGRAM DWARF_CLOUDSC
 
+
+#ifdef CLOUDSC_GPU_SCC_FIELD
+TYPE(CLOUDSC_FIELD_STATE) :: FIELD_STATE
```
Yes, although FIELD_STATE has an additional option. However, the call to verify now aliases nicely, and I moved the invocation to the back to keep with the overall order.

mlange05 commented:
Ok, thanks for the reviews. I think I've addressed all point, but please check. I've also had to rebase, but the changes since last review are in the last 3 commits.

reuterbal commented:
Looks all good now, many thanks! GTG
